ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+        if bad_prefix:
+            logger.info("Client started request with non json data: %r", bad_prefix)
+            await write_error(socket, f"Cannot parse json: {bad_prefix}")
+            continue
+
+        try:
+            request, offset = decoder.raw_decode(raw_request)
+        except json.JSONDecodeError:
+            # invalid json request, keep reading data until a valid json is formed
+            if raw_request:
+                logger.debug(
+                    "Invalid JSON, waiting for rest of message: %r", raw_request,
+                )
+            else:
+                await trio.sleep(0.01)
+            continue
+
+        # TODO: more efficient algorithm can be used here by
+        # manipulating the buffer such that we can seek back to the
+        # correct position for *new* data to come in.
+        buffer.seek(0)
+        buffer.write(raw_request[offset:])
+        buffer.truncate()
+
+        break
+
+    return request
+
+
+async def write_error(socket: trio.socket.SocketType, message: str) -> None:
+    json_error = json.dumps({"error": message})
+    await socket.send(json_error.encode("utf8"))
+
+
+def validate_request(request: Mapping[Any, Any]) -> None:
+    try:
+        version = request["jsonrpc"]
+    except KeyError as err:
+        raise ValidationError("Missing 'jsonrpc' key") from err
+    else:
+        if version != "2.0":
+            raise ValidationError(f"Invalid version: {version}")
+
+    if "method" not in request:
+        raise ValidationError("Missing 'method' key")
+    if "params" in request:
+        if not isinstance(request["params"], list):
+            raise ValidationError("Missing 'method' key")
+
+
+def generate_response(
+    request: RPCRequest, result: Any, error: Optional[str]
+) -> RPCResponse:
+    response = RPCResponse(
+        id=request.get("id", -1), jsonrpc=request.get("jsonrpc", "2.0"),
+    )
+
+    if result is None and error is None:
+        raise TypeError("Must supply either result or error for JSON-RPC response")
+    if result is not None and error is not None:
+        raise TypeError(
+            "Must not supply both a result and an error for JSON-RPC response"
+        )
+    elif result is not None:
+        response["result"] = result
+    elif error is not None:
+        response["error"] = str(error)
+    else:
+        raise Exception("Unreachable code path")
+
+    return response
+
+
+TParams = TypeVar("TParams")
+TResult = TypeVar("TResult")
+
+
+class RPCError(Exception):
+    ...
+
+
+class RPCHandler(RPCHandlerAPI, Generic[TParams, TResult]):
+    async def __call__(self, request: RPCRequest) -> RPCResponse:
+        try:
+            params = self.extract_params(request)
+            result = await self.do_call(params)
+        except RPCError as err:
+            return self.generate_error_response(request, str(err))
+        else:
+            return self.generate_success_response(request, result)
+
+    @abstractmethod
```

Comment:
*nit*: If someone reads through this class this will already be pretty clear, it's just 6 lines up. But to be more explicit, it could be useful to include a docstring here explaining that you should throw `RPCError` if you want your error to be returned to the caller.